202509 rechnungsfreigabe

jbueren opened the pull request:
Der simplere Teil des ukd-PR3 Pull Requests:

- Es gibt ein neues Feld Einkäufer für Lieferanten, die im Lieferanten-Workflow
  bei den Belegen gesetzt werden können.

- Konfigurierbares Feature in der Mandatenkonfiguration:
  Lieferantenrechnungen können erst nach Freigabe SEPA überwiesen werden.

kivijan committed:
Kreditorenbuchungen: Zahlungen freigeben DB und ROSE

kivijan committed:
db-relation Zahlung freigeben Kreditorenbuchung

kivijan committed:
ActionBar Zahlung genehmigen für Kreditoren und EK-Rechnungen

kivijan committed:
Zahlung genehmigt am Tag, als optionales Anhakfeld im Bericht

kivijan committed:
Zahlung genehmigt am und von als Infofeld in Buchungsmasken anzeigen

kivijan committed:
SEPA:retrieve_open_invoice: Parameter nur freigegebene Rechnungen auslesen

kivijan committed:
sepa: Aufruf mit Parameter only_approved und Info zur Freigabe für 1. Liste

kivijan committed:
Template SEPA: Info zur Freigabe immer anzeigen. Sperren falls MandKonf

kivijan committed:
SEPA Lauf für Eingangsrechnungen nur nach Freigabe (Mand. Konfiguration)

kivijan committed:
lokalisierung vereinheitlichen

kivijan committed:
locales

kivijan committed:
Einkäufer für Lieferantenstamm hinzugefügt

kivijan committed:
Einkäufer für Einkaufsbelege in der DB hinzugefügt

kivijan committed:
Lieferant Einkäufer vorbelegen. Fehlerhafte end_customer_id aus rose raus

kivijan committed:
Einkäufer gespiegelt zu Verkäufer für Order-Controller hinzugefügt

kivijan committed:
Order-Controller: der default für buyer_id wird in Model gesetzt

kivijan committed:
Einkäufer auch in DeliverOrder-Controller hinzugefügt

kivijan committed:
rose Order.pm korrekt generieren (ohne Projektcode)

kivijan committed:
Einkaufsrechnung Verkäufer erfassen und vorbelegen

kivijan committed:
Workflow Lieferantenauftrag -> Rechnung Einkäufer nicht vergessen

kivijan committed:
Reklamationen haben auch einen Einkäufer

kivijan committed:
Lieferscheine dürfen Einkäufer übernehmen. Ausnahme für sales_inv in Test

kivijan committed:
Einkäufer auch für Kreditorenbuchung (WF und change_vc i.O.)

Zusätzlich Bearbeiter anzeigen und editierbar gemacht

kivijan committed:
set initial Einkäufer bei Kreditorenbuchung

kivijan committed:
SQL-syntax fehler ;-(

kivijan committed:
neues Rose Object in ALL

kivijan committed:
locales

kiviniklas reviewed (changes requested):
Gut gefällt mir, dass nun einkaufsseitig der Buyer symmetrisch zum Salesman verkaufsseitig eingeführt wird, und dass das Datenbankfeld dafür eine klar unterscheidbare Bezeichnung trägt.

Einige Kleinigkeiten am Code habe ich angemerkt.

Changelog ziehe ich selbst gleich aus dem #569 herüber, soweit zutreffend.

Funktionaler Test folgt direkt im Anschluss.

kiviniklas committed:
changelog: Feld Einkäufer, Freigabe EK-Rechnungen erklärt

kiviniklas committed:
Kreditoren- und EK-Rechnungen: Zahlung freigeben auch leer anzeigen

Wenn geprüft werden soll, ob eine Rechnung noch nicht freigegeben
ist, ist es klarer, wenn die Feldbezeichnung "Zahlung freigegeben"
angezeigt wird und der Feldinhalt leer gezeigt wird.

Die Prüfung, ob die Freigabe erfolge, bevor auf emloyee.name und
itime_as_date zugriffen wird, benötigen wir in den Templates nicht.
Daher ist diese nun entfernt.

kiviniklas commented:
@jbueren Für den Standard überlasse ich dir die o.g. Änderungsvorschläge, wenn du einverstanden bist.

bblessmann reviewed (commented):
Erstmal nur kurz drübergeschaut und dokumentiert, was mit hier im Zug so aufgefallen ist.

kivijan committed:
Einkäufer immer setzen, analog zu Verkäufer

kivijan committed:
Einkäufer auch anzeigen, falls der Benutzer gelöscht wird

kivijan committed:
Refactor: payment_approvals (Pl.) statt payment_approved

kivijan committed:
Bericht Ek-Rechnungen Einkäufer als Spalte anzeigen

kiviniklas committed:
Bericht Ek-Rechnungen Einkäufer als Spalte anzeigen

kiviniklas commented:
Was noch nicht funktioniert: buyer_id in der ap Tabelle wird beim Buchen einer Rechnung nicht geschrieben. (Wert ist immer NULL)

kiviniklas committed:
fix: Lieferant Einkäufer vorbelegen.

Sonst Fehler beim Aufrufen von Kunden in der Oberfläche

kivijan committed:
Einkäufer auch bei Einkaufsrechnung speichern

kivijan committed:
kosmetik: bessere lesbarkeit if/else

jbueren commented on SL/Model/Record.pm:
```diff
@@ -58,7 +61,7 @@ sub update_after_customer_vendor_change {
       $record->billing_address_id($address ? $address->id : undef);
     }
-  }
+  } else { die "invalid state"; }
```
Gut einverstanden, es war auch vorher logisch andersrum.
Den else Zweig lass ich aber (auch wenn der __nie eintrifft).

jbueren commented on sql/Pg-upgrade2/ap_add_employee_approved.sql:
```diff
@@ -0,0 +1,14 @@
+-- @tag: ap_add_employee_approved
+-- @description: Spalte für Genehmigt bei EK-Rechnungen
+-- @depends: release_3_9_1
+CREATE TABLE payment_approved (
```
Besser jetzt refactoren als später. Geändert

bblessmann reviewed (approved):
Hi Jan,

warum ist die Beziehung zu payment_approvals eine 1-n-Beziehung? Kann es mehrere Freigaben (von verschiedenen Benutzern) geben? Das geht im Moment an der Oberfläche nicht, oder?
Aber wenn dann doch zwei Freigaben zu einer Rechnung existieren, dann stimmt der Rechnungsbericht in den Summern nicht, da durch das LEFT JOIN dann mehrere Zeilen zu einer Rechnung auftauchen können.
Auch liefert $i->find_payment_approvals (übrigens geht auch $i->payment_approvals) ein Array, Du verwendest es zumindest in den Templates als Scalar/Objekt.
Evtl. ist eine 1-1-Beziehung besser?

Bei mir funktioniert die Anzeige von "Zahlung freigegeben" im Sepa-Lauf nicht. Spalte ist immer leer. Und sollen nur freigegebene angezeigt werden, wenn das Feature an ist? Das klappt bei mir auch nicht. Oder werden alle angezeigt, sind aber nicht auswählbar? Dann wäre ein Hinweis nicht schlecht.

Bei der Einkaufsrechnung wird der Einkäufer nicht aus den Stammdaten vorbelegt, wenn der Lieferant gewechselt wird.

Soweit erstmal. Ist ja doch noch groß der PR.

kivijan committed:
dämliche Macke, wirklich den Einkäufer der Rechnung speichern

kiviniklas commented on bin/mozilla/ir.pl:
```diff
@@ -703,6 +713,14 @@ sub block_or_unblock_sepa_transfer {
+
+  $::form->redirect(t8('Payment Approved'));
+}
+
```
> Eine Funktion höher wird auch nur auf das recht ap_transactions geprüft.

Mir geht es um folgendes:
Der versierte Nutzer kann einen HTTP Request zusammendengeln, welcher die bisherige Zahlungsfreigabe des Beleges überschreibt durch eine Neue Zahlungsfreigabe.
Das ist jetzt nicht ganz so schlimm.

Grundsätzlich finde ich, unser Backend sollte nur genau das erlauben, was auch bspw. in der ActionBar als legitime Option angezeigt wird. Sonst ist ein Audit des Codes schwieriger, weil man die Absichten des ursprünglichen Entwicklers dann erraten muss, sofern nicht dokumentiert.

Das ist jetzt aber super detailliert hier. Ist auch so ok.

kiviniklas commented:
Steht fest, ob es zu einem Beleg mehrere Zahlungsfreigaben geben können soll, oder nur maximal eine? Wenn das klar ist, baue ich gerne class="data" ins Template.

kivijan committed:
Revert "Einkäufer immer setzen, analog zu Verkäufer"

This reverts commit f9cfec731094eb9b723fba23326a65566ca0e328.

kivijan committed:
Einkäufer ist Pflichtfeld in Kreditorenbuchung

bblessmann commented:
Wird als #823 weitergeführt. Dort sind die Kommentare von hier weitgehend berücksichtig; rebased, gefixt und etwas refactored.
